spec/lib/discourse_spec.rb

```diff
@@ -340,6 +340,32 @@ def call(ex, ctx)
       Sidekiq.error_handlers.delete(logger)
     end
 
+    it "should collect job exception stats" do
+      Discourse.reset_job_exception_stats!
```

Comment:
I think we need to ensure that the stat is reset after each test and not before. Currently, this test will leak state after it runs.

Reply:
sure I will reset both before and after, I am not sure I want to open the door for an arbitrary flaky spec

Reply:
I think it is better to only reset after because resetting before will hide a state leak. I rather we have visibility than none. Alternatively, we can just reset it after the end of each test run regardless since the reset is so cheap anyway.

Reply:
I am not sure what visibility is buying us though, it is a lot of busywork just to save one call to resetting stats. 

Any place that results in a job failure now needs to worry about resetting, what value does catching this add? 

I can just test for delta if you prefer then it no longer resets at all.

Reply:
> Any place that results in a job failure now needs to worry about resetting, what value does catching this add?

Wouldn't it be better to just add a reset after each tests regardless of whether there is a job failure? Whenever each test case runs, we should be confident that the starting state is always the same. The `before` hook here indicates to me that we lack that confidence which is problematic IMO because it means the currently changes can potentially allow a state leak.

Reply:
I agree state leaks are a bad thing, but this is simply leaking statistics, its a bit of a wash to track the leaks down cause they do not matter in any material way. 

If we really wanted to solve this super cleanly I guess we would "disable stats" globally and only enable them strategically for the specific test, but it is a fair amount of ceremony for very little gain.

Reply:
> its a bit of a wash to track the leaks down cause they do not matter in any material way.

They may not matter at this point in time but we can't assume the same for the future when there may be multiple test cases for this statistics we're tracking. My opinion here is to not set a trap for someone in the future where they introduce a flaky test because they forgot to run `Discourse.reset_job_exception_stats!` before their assertion.

> If we really wanted to solve this super cleanly I guess we would "disable stats" globally and only enable them strategically for the specific test, but it is a fair amount of ceremony for very little gain.

This is a good point and got me thinking. Should this even be enabled by default? We're kind of doing wasteful work here on sites that do not need the stats.